src/workerd/api/node/crypto.c++

```diff
@@ -1141,4 +1141,176 @@ jsg::Optional<CryptoImpl::CipherInfo> CryptoImpl::getCipherInfo(
+#pragma region ECDH
+
+namespace {
+ncrypto::ECPointPointer bufferToPoint(const EC_GROUP* group, jsg::BufferSource& buf) {
+  JSG_REQUIRE(buf.size() <= INT32_MAX, Error, "buffer is too big");
+
+  auto pub = ncrypto::ECPointPointer::New(group);
+  JSG_REQUIRE(pub, Error, "Failed to allocate EC_POINT for a public key");
+
+  ncrypto::Buffer<const unsigned char> buffer{
+    .data = buf.asArrayPtr().begin(),
+    .len = buf.size(),
+  };
+
+  JSG_REQUIRE(pub.setFromBuffer(buffer, group), Error, "Failed to set point");
+  return pub;
+}
+
+point_conversion_form_t getFormat(kj::StringPtr format) {
+  if (format == "compressed"_kj) return POINT_CONVERSION_COMPRESSED;
+  if (format == "uncompressed"_kj) return POINT_CONVERSION_UNCOMPRESSED;
+  if (format == "hybrid"_kj) return POINT_CONVERSION_HYBRID;
+  JSG_FAIL_REQUIRE(Error, "Invalid ECDH public key format");
+}
+
+jsg::BufferSource ecPointToBuffer(
+    jsg::Lock& js, const EC_GROUP* group, const EC_POINT* point, point_conversion_form_t form) {
+  size_t len = EC_POINT_point2oct(group, point, form, nullptr, 0, nullptr);
+  JSG_REQUIRE(len != 0, Error, "Failed to get public key length");
+
+  auto backing = jsg::BackingStore::alloc<v8::ArrayBuffer>(js, len);
+
+  len =
+      EC_POINT_point2oct(group, point, form, backing.asArrayPtr().begin(), backing.size(), nullptr);
+  JSG_REQUIRE(len != 0, Error, "Failed to get public key");
+
+  return jsg::BufferSource(js, kj::mv(backing));
+}
+
+bool isKeyValidForCurve(const EC_GROUP* group, const ncrypto::BignumPointer& private_key) {
+  // Private keys must be in the range [1, n-1].
+  // Ref: Section 3.2.1 - http://www.secg.org/sec1-v2.pdf
+  if (private_key < ncrypto::BignumPointer::One()) {
+    return false;
+  }
+  auto order = ncrypto::BignumPointer::New();
+  JSG_REQUIRE(order, Error, "Internal failure when checking ECDH key");
+  return EC_GROUP_get_order(group, order.get(), nullptr) && private_key < order;
+}
+}  // namespace
+
+CryptoImpl::ECDHHandle::ECDHHandle(ncrypto::ECKeyPointer key)
+    : key_(kj::mv(key)),
+      group_(key_.getGroup()) {}
+
+jsg::Ref<CryptoImpl::ECDHHandle> CryptoImpl::ECDHHandle::constructor(
+    jsg::Lock& js, kj::String curveName) {
+
+  int nid = OBJ_sn2nid(curveName.begin());
+  JSG_REQUIRE(nid != NID_undef, Error, "Invalid curve");
+
+  auto key = ncrypto::ECKeyPointer::NewByCurveName(nid);
+  JSG_REQUIRE(key, Error, "Failed to create key using named curve");
+
+  return jsg::alloc<CryptoImpl::ECDHHandle>(kj::mv(key));
+}
+
+jsg::BufferSource CryptoImpl::ECDHHandle::computeSecret(
+    jsg::Lock& js, jsg::BufferSource otherPublicKey) {
+
+  ncrypto::ClearErrorOnReturn clear_error_on_return;
+
+  JSG_REQUIRE(key_.checkKey(), Error, "Invalid keypair");
+
+  auto pub = bufferToPoint(group_, otherPublicKey);
+  JSG_REQUIRE(pub, Error, "Invalid to set ECDH public key");
+
+  int field_size = EC_GROUP_get_degree(group_);
+  size_t out_len = (field_size + 7) / 8;
```

Comment:
How come no static cast is required here? Field_size is int, and casting it into size_t would remove the unsigned attribute.

Reply:
converting from signed to unsigned here is generally safe. Plus this is copied directly from node.js